### DISCUSSION TITLE
Ory Hydra using GCP Cloud Identity for authentication

### DISCUSSION THREAD

**Discussion by anupammishra-sada** [ory/hydra#3580] [Q&A] [answered]
Hi Guys, I am new to Ory and looking for some documentation in which I can do the authentication via GCP cloud Identity. Can someone point me to the documentation or suggest the steps required.

**Comment by vinckr** [comment]
Hello @anupammishra-sada,

have you seen this document: https://www.ory.sh/docs/kratos/social-signin/google ?

**Reply by anupammishra-sada** [reply]
Hi @vinckr yes I did. What I am not sure about is how to configure local Ory Hydra instance to use GCP Cloud Identity as the IDP. I am unable to understand how to configuration done in the Ory console for a project maps int he Ory Hydra code. Can you guide me through it or provide any documentation.

**Reply by vinckr** [reply]
This is not possible using just Ory Hydra AFAIK. 
You need something like [Ory Kratos](https://github.com/ory/kratos) that can act as the OIDC client. Hydra is an OIDC server. 

You can export configuration from Ory Network and use it in the respective Ory services (Kratos, Hydra, etc...): https://www.ory.sh/docs/guides/cli/config-with-cli

**Comment by anupammishra-sada** [comment]
Sorry @vinckr I wasn't clear with the ask probably and still learning Ory. I know that Ory Hydra can be integrated with Ory Kratos but isn't that necessitates me to maintain all my user identities in Ory Kratos. If yes, that is where my struggle is to use google cloud identity based users for authentication and use Hydra for authorization. ? So instead of maintaining and verifying the identity in Kratos it is done with GCP cloud identity. It might not be available out of the box, but can it be configured if required. ?

**Reply by vinckr** [reply]
Hey @anupammishra-sada 

You can add Google as OIDC provider in Ory Kratos: https://www.ory.sh/docs/kratos/social-signin/google 

Maybe we can take a step back and you describe:

- what is your current setup
- why do you want to change it
- what is your requirements when you change it
- how should the setup look after the change

that would help give a better recommendation!

**Reply by anupammishra-sada** [reply]
Hi @vinckr we are planning to use Ory Selfhosting. PFB the answer against each. The basic use case is given below:

We have a frontend application with a login screen, we would like to use that screen to register and sign in the users. The users will be maintained in the identity server. The application need JWT to be generated after successful login. 

Sometime later there has to be a mechanism to expose our rest API only to the authorized services

**what is your current setup**

We are starting our development and considered Ory to be used.  In this we are planning to use Ory Hydra in combination with an identity server. Hence the question can the identity server be anything other than Kratos.

**why do you want to change it**

Most of our identities are maintained in GCP cloud Identity. So if Ory Hydra can be integrated with Google cloud identity we will not have to get the identities to Ory Kratos.

**what is your requirements when you change it**

Ory Hydra with Google cloud Identity as Identity server.

**Reply by vinckr** [reply]
>can the identity server be anything other than Kratos.

Yes definitely!
You can use anything that works with OAuth2. Ory Kratos is of course specifically developed to integrate with Hydra. It works out of the box on the managed service and you can check out a self-hosting example [here](https://github.com/ory/examples/tree/master/kratos-hydra) or on [our docs](https://www.ory.sh/docs/welcome)
See this document on OAuth2 tokens and JWT: https://www.ory.sh/docs/oauth2-oidc/jwt-access-token

You still need to manage the users in your identity server but if that is already set up should be fine. You also need to code the frontend application to handle the sign in etc.

**Reply by anupammishra-sada** [reply]
Hi @vinckr Thanks a lot. Another question as a follow up, say incase I have multiple IDP. So take an example I have multiple users a@123.com, b@xyz.com these Id's are maintained in there respective IDP's. 
1. Can I connect Ory Hydra to multiple IDP.?
2. When they login to my application based on the **mailID** can I do the authentication from the respective IDP. ?

**Reply by vinckr** [reply]
1. Yes, Ory Hydra can be connected to multiple Identity Providers (IDPs). 

Ory Hydra itself is not an identity management system, but it can connect to your existing identity management system, which could be any IDP you have in place. This could be a MySQL+PHP login service, a Federated SAML SSO, or any other system you use to manage identities.

2. When a user tries to access your application, Ory Hydra can issue access, refresh, and ID tokens in a secure and OpenID Certified manner, connecting to the respective IDP based on the user's credentials. However, the specifics of how to route authentication requests based on the email domain (like a@123.com to IDP1, b@xyz.com to IDP2) would depend on the implementation details of your login provider. The login provider would need to be set up to handle this routing logic.
See here for example how you would do it with [Ory Kratos](https://www.ory.sh/docs/kratos/social-signin/overview).

**Reply by anupammishra-sada** [reply]
Hi @vinckr , Does it mean, in case of Federated Identity Ory Kratos act as a Service Provider and can connect to multiple IDP's including SAML based IDP. Can you please point me to a documentation around how this can be achieved in a self hosted Ory Kratos instance.

**Reply by vinckr** [reply, marked as answer]
See here for OIDC: https://www.ory.sh/docs/self-hosted/kratos/configuration/oidc 
Out of the box SAML is work in progress, but you can use something else for this functionality in the meantime.

**Reply by anupammishra-sada** [reply]
Thanks a lot @vinckr . Is there a feature request link to SAML feature which is WIP. ? And what do you mean by something else say if the IDP only supports SAML. ?

**Reply by vinckr** [reply]
Hey @anupammishra-sada 
You can see the PR here: https://github.com/ory/kratos/pull/2653 
And there are SAML bridge projects that you can use with like [jackson](https://github.com/boxyhq/jackson).

If you are interested in getting SAML into Ory, I would suggest to reach out to support@ory.sh
